networkx/algorithms/community/quality.py

```diff
@@ -276,45 +292,41 @@ def modularity(G, communities, weight="weight"):
     >>> import networkx.algorithms.community as nx_comm
     >>> G = nx.barbell_graph(3, 0)
     >>> nx_comm.modularity(G, [{0, 1, 2}, {3, 4, 5}])
-    0.35714285714285704
+    0.35714285714285715
     >>> nx_comm.modularity(G, nx_comm.label_propagation_communities(G))
-    0.35714285714285704
+    0.35714285714285715
 
     References
     ----------
     .. [1] M. E. J. Newman *Networks: An Introduction*, page 224.
        Oxford University Press, 2011.
-
+    .. [2] Clauset, Aaron, Mark EJ Newman, and Cristopher Moore.
+    "Finding community structure in very large networks."
+    Physical review E 70.6 (2004). <https://arxiv.org/abs/cond-mat/0408187>
     """
     if not isinstance(communities, list):
         communities = list(communities)
     if not is_partition(G, communities):
         raise NotAPartition(G, communities)
 
-    multigraph = G.is_multigraph()
     directed = G.is_directed()
-    m = G.size(weight=weight)
     if directed:
         out_degree = dict(G.out_degree(weight=weight))
         in_degree = dict(G.in_degree(weight=weight))
-        norm = 1 / m
+        m = sum(out_degree.values())
+        norm = 1 / m ** 2
     else:
-        out_degree = dict(G.degree(weight=weight))
-        in_degree = out_degree
-        norm = 1 / (2 * m)
-
-    def val(u, v):
-        try:
-            if multigraph:
-                w = sum(d.get(weight, 1) for k, d in G[u][v].items())
-            else:
-                w = G[u][v].get(weight, 1)
-        except KeyError:
-            w = 0
-        # Double count self-loops if the graph is undirected.
-        if u == v and not directed:
-            w *= 2
-        return w - in_degree[u] * out_degree[v] * norm
-
-    Q = sum(val(u, v) for c in communities for u, v in product(c, repeat=2))
-    return Q * norm
+        out_degree = in_degree = dict(G.degree(weight=weight))
+        m = sum(out_degree.values()) / 2
+        norm = 1 / (2 * m) ** 2
```

Comment:
Why the division by two immediately followed by `2 * m` here? Don't these factors of two cancel out? And if so, it seems it's unnecessary to have these two lines inside the conditional. Sorry if I'm missing something obvious!

Reply:
No worries. `m` is also used in [line 331](https://github.com/networkx/networkx/pull/4103/files#diff-0b456323a189caba1c86fb3003b7d602R331) in the return statement of the community contribution. Depending on whether `G` is directed, we need a different value for `m` there. That's why the calculation is in the conditional and in a separate statement from the norm. At least from my perspective, I don't see how I could simplify the calculation further.

Reply:
Ah, I see. Then maybe something like:

```python
if directed:
    out_degree = dict(G.out_degree(weight=weight))
    in_degree = dict(G.in_degree(weight=weight))
else:
    out_degree = in_degree = dict(G.degree(weight=weight))

deg_sum = sum(out_degree.values())
norm = 1 / deg_sum ** 2
m = deg_sum if directed else deg_sum / 2
```

It doesn't really matter for performance or anything, but the `x = 2 * y / 2` pattern stands out to me as not ideal for readability. It's not a blocker for me, so feel free to stick with whatever makes the most sense to you!